### PR TITLE
Ensured variables fills automatically in Make date dialog

### PR DESCRIPTION
**Pull request by Vitalis95**
Fixes partially #8213
Fixes partially part c) of the issue
@rdstern , @lloyddewit , Ensured variables fills automatically in all the buttons, swapped `single column` and `Year Month Day` and  in the 3 variable case, ensured  it goes nicely from year to month to day

### CODE REVIEW

**Review by rdstern — DISMISSED** [review]
@Vitalis95 this looks a really good simplification for the users.  If @lloyddewit is ok I would like to merge this, before waiting for anything else.

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Vitalis95 Thank you for the changes.
I didn't review the code changes yet but I noticed that this PR includes 3 extra packages. These seem to be older versions of packages already included in the R-Instat repo. Please could you remove these packages from this PR?
(let me know if you have any questions).
Thanks

P.S. These are the packages I have:

image

**Comment by Vitalis95** [comment]
@lloyddewit , check it now

File: instat/dlgMakeDate.vb [hunk 1]
```diff
@@ -15,17 +15,26 @@
...
+Imports System.Text.RegularExpressions
+Imports RDotNet
 Public Class dlgMakeDate
     Public clsPaste As New RFunction
     Public bFirstLoad As Boolean = True
     Private bReset As Boolean = True
     Private bUseSelectedColumn As Boolean = False
+    Dim clsTypesFunction As New RFunction
+    Dim lstReceivers As New List(Of ucrReceiverSingle)
+    Dim lstRecognisedTypes As New List(Of KeyValuePair(Of String, List(Of String)))
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:27]: better to use a dictionary type?
```suggestion
   Dim dctRecognisedTypes As New Dictionary(Of String, List(Of String))
```

File: instat/dlgMakeDate.vb [hunk 2]
```diff
@@ -406,6 +455,58 @@ Public Class dlgMakeDate
...
+
+        For Each kvpTemp As KeyValuePair(Of String, List(Of String)) In lstRecognisedTypes
+            If kvpTemp.Key = strVariable Then
+                lstValues = kvpTemp.Value
+                Exit For
+            End If
+        Next
+        Return lstValues
+    End Function
+
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:509]: delete this function? (see suggestion above)
```suggestion
```

File: instat/dlgMakeDate.vb [hunk 3]
```diff
@@ -406,6 +455,58 @@ Public Class dlgMakeDate
...
+        Dim ucrCurrentReceiver As ucrReceiver
+        Dim bFound As Boolean = False
+
+        ucrCurrentReceiver = ucrSelectorMakeDate.CurrentReceiver
+
+        For Each ucrTempReceiver As ucrReceiver In lstReceivers
+            ucrTempReceiver.SetMeAsReceiver()
+            lstRecognisedValues = GetRecognisedValues(ucrTempReceiver.Tag)
+
+            If lstRecognisedValues.Count > 0 Then
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:476]: Use dictionary and follow 'happy path keeps left' principle (to minimize indents)?
```suggestion
            Dim lstRecognisedValues As List(Of String) = dctRecognisedTypes.Item(ucrTempReceiver.Tag)
            If IsNothing(lstRecognisedValues) Then
                Continue For
            End If
```

File: instat/dlgMakeDate.vb [hunk 4]
```diff
@@ -406,6 +455,58 @@ Public Class dlgMakeDate
...
+                        If Regex.Replace(lviTempVariable.Text.ToLower(), "[^\w]|_", String.Empty).Contains(strValue) Then
+                            ucrTempReceiver.Add(lviTempVariable.Text, ucrSelectorMakeDate.ucrAvailableDataFrames.cboAvailableDataFrames.Text)
+                            bFound = True
+                            Exit For
+                        End If
+                    Next
+                    If bFound Then
+                        bFound = False
+                        Exit For
+                    End If
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:488]: more readable to write on one line?
```suggestion
                If bFound Then Exit For
```

File: instat/dlgMakeDate.vb [hunk 5]
```diff
@@ -15,17 +15,26 @@
...
     Private strSelectedColumn As String = ""
     Private strSelectedDataFrame As String = ""
     Private clsDateFunction, clsMakeYearDay, clsHelp, clsMakeYearMonthDay, clsDefaultDate, clsRDefaultDate As New RFunction
     Private clsGregorianDefault, clsJulianDateDefault, clsAsCharacterFunction, clsDummyFunction As New RFunction
+    Private clsDefaultFunction As New RFunction
+    Private clsConcFunction As New RFunction
     Private clsDivisionOperator, clsMultiplicationOperator As New ROperator
 
+
+
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:37]: In the rest of the code, we normally only leave one blank line between sections. We should be consistent.
```suggestion
```

File: instat/dlgMakeDate.vb [hunk 6]
```diff
@@ -48,6 +57,25 @@ Public Class dlgMakeDate
         Dim dctMonthTwoItems As New Dictionary(Of String, String)
         Dim dctDateFormat As New Dictionary(Of String, String)
 
+        Dim kvpDate As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("date", {"date", "record"}.ToList())
+        Dim kvpYear As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("year", {"year", "yr"}.ToList())
+        Dim kvpMonth As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("month", {"month", "mon", "mois"}.ToList())
+        Dim kvpDay As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("day", {"day", "jour"}.ToList())
+        Dim kvpDOY As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("doy", {"doy", "doy_366"}.ToList())
+
+
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:66]:
```suggestion

```

File: instat/dlgMakeDate.vb [hunk 7]
```diff
@@ -48,6 +57,25 @@ Public Class dlgMakeDate
...
+        Dim kvpYear As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("year", {"year", "yr"}.ToList())
+        Dim kvpMonth As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("month", {"month", "mon", "mois"}.ToList())
+        Dim kvpDay As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("day", {"day", "jour"}.ToList())
+        Dim kvpDOY As KeyValuePair(Of String, List(Of String)) = New KeyValuePair(Of String, List(Of String))("doy", {"doy", "doy_366"}.ToList())
+
+
+        lstRecognisedTypes.AddRange({kvpYear, kvpMonth, kvpDay, kvpDOY, kvpDate})
+        lstReceivers.AddRange({ucrReceiverDayThree, ucrReceiverDayTwo, ucrReceiverForDate, ucrReceiverMonthThree, ucrReceiverYearThree, ucrReceiverYearTwo})
+
+
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:70]:
```suggestion
```

File: instat/dlgMakeDate.vb [hunk 8]
```diff
@@ -296,6 +331,10 @@ Public Class dlgMakeDate
         clsMultiplicationOperator = New ROperator
         clsDummyFunction = New RFunction
 
+        clsDefaultFunction = New RFunction
+        clsConcFunction = New RFunction
+
+
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:337]:
```suggestion
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Comment by Vitalis95** [comment]
@lloyddewit , have a look at it

File: instat/dlgMakeDate.vb [hunk 9]
```diff
@@ -406,6 +447,49 @@ Public Class dlgMakeDate
...
+    Private Sub AutoFillReceivers()
+        Dim ucrCurrentReceiver As ucrReceiver = ucrSelectorMakeDate.CurrentReceiver
+        Dim lstRecognisedValues As List(Of String)
+
+        For Each ucrTempReceiver As ucrReceiver In lstReceivers
+            Dim bFound As Boolean = False
+            ucrTempReceiver.SetMeAsReceiver()
+            lstRecognisedValues = GetRecognisedValues(ucrTempReceiver.Tag)
+            If IsNothing(lstRecognisedValues) Then
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:465]: would something like this work?
(we could then delete the function below)
```suggestion
        Dim lstRecognisedValues As List(Of String) = New List(Of String)

        For Each ucrTempReceiver As ucrReceiver In lstReceivers
            Dim bFound As Boolean = False
            ucrTempReceiver.SetMeAsReceiver()
            If Not dctRecognisedTypes.TryGetValue(ucrTempReceiver.Tag, lstRecognisedValues) Then
```

File: instat/dlgMakeDate.vb [hunk 10]
```diff
@@ -406,6 +447,49 @@ Public Class dlgMakeDate
...
+                If bFound Then Exit For
+            Next
+        Next
+
+        If ucrCurrentReceiver IsNot Nothing Then
+            ucrCurrentReceiver.SetMeAsReceiver()
+        End If
+    End Sub
+
+    Private Function GetRecognisedValues(strVariable As String) As List(Of String)
```
Comment (lloyddewit) [inline instat/dlgMakeDate.vb:485]: delete this function?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
looks good, just one suggestion

**Comment by Vitalis95** [comment]
@lloyddewit , it is fine now

**Review by lloyddewit — APPROVED** [review]

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 I am getting an error in some cases:

image

I got this with the dodoma.rds data from the library.  

When I run simply, it fills the dialogue nicely and runs with no problem.  It gives me date1, because date is already there.

I now changed the names, so it didn't recognise them - so year into a, month into b and day into c.  Now I have to fill the year, month and day manually.  I now get this error:

image

I notice it generated an odd command namely:

```
# Code generated by the dialog
date2 <- data_book$make_date_yearmonthday(data_name="dodoma", year_format="%Y", month_format="%m")
data_book$add_columns_to_data(data_name="dodoma", col_name="date2", col_data=date2, before=FALSE, adjacent_column="a")
```
When I return to the dialogue, make no changes, then it runs fine and generates the correct code as follows:

```
# Code generated by the dialog
date2 <- data_book$make_date_yearmonthday(data_name="dodoma", month="b", year="a", day="c", year_format="%Y", month_format="%m")
data_book$add_columns_to_data(data_name="dodoma", col_name="date2", col_data=date2, before=TRUE, adjacent_column="a")

```

**Comment by Vitalis95** [comment]
@rdstern , can you have look at it

**Review by rdstern — CHANGES_REQUESTED** [review]
@Vitalis95 it now seems to work almost fine.  I tried again starting with 3 variables it doesn't recognise.  It now works in that there is no error.  It gives this code:


```
# Code generated by the dialog
date <- data_book$make_date_yearmonthday(data_name="Uzgen", day="D.D.D.", month="D.D.N.N.N.", year="D3D.D.", year_format="%Y", month_format="%m")
data_book$add_columns_to_data(data_name="Uzgen", col_name="date", col_data=date, before=FALSE, adjacent_column="D3D.D.")
```
When I run again it has before=TRUE, which I prefer!

**Review by rdstern — COMMENTED** [review]
@Vitalis95 I am keen to get these changes merged.  Can you make the fix, so that Before - TRUE each time?
Here are my 2 runs with the Dodoma data (after changing the name to `a` for the year:

image

**Comment by rdstern** [comment]
@Vitalis95 will you soon be able to make this (small?) change, or are you stuck?  I am keen for this improvement to be merged as soon as possible.

**Comment by Vitalis95** [comment]
> @Vitalis95 will you soon be able to make this (small?) change, or are you stuck? I am keen for this improvement to be merged as soon as possible.

@rdstern , I am stuck on that, the code is Ok but it is not working,  We discussed with @N-thony and he promised to look into it,  I suspect it might be the changes done by Patrick on the Save controls

**Comment by N-thony** [comment]
> > @Vitalis95 will you soon be able to make this (small?) change, or are you stuck? I am keen for this improvement to be merged as soon as possible.
> 
> @rdstern , I am stuck on that, the code is Ok but it is not working, We discussed with @N-thony and he promised to look into it, I suspect it might be the changes done by Patrick on the Save controls

I will look into it meanwhile this can be merged.

**Comment by lloyddewit** [comment]
@rdstern Shall we merge this PR for this week's release or would you prefer to keep this PR open for other changes?
thanks

**Review by lloyddewit — APPROVED** [review]

**Pull request merged**